### PR TITLE
[mlir][spirv] Add folding for [I|Logical][Not]Equal

### PR DESCRIPTION
**Pull request by inbelic**

### CODE REVIEW

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 1]
```diff
@@ -309,19 +309,62 @@ OpFoldResult spirv::LogicalAndOp::fold(FoldAdaptor adaptor) {
...
+  if (getOperand1() == getOperand2()) {
+    auto type = getType();
+    if (isa<IntegerType>(type)) {
+      return BoolAttr::get(getContext(), true);
+    }
+    if (isa<VectorType>(type)) {
+      auto vtType = cast<ShapedType>(type);
+      auto element = BoolAttr::get(getContext(), true);
+      return DenseElementsAttr::get(vtType, element);
+    }
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:328]: nit:
```suggestion
    auto trueAttr = BoolAttr::get(getContext(), true);
    if (isa<IntegerType>(getType())) {
      return tureAttr;
    }
    if (auto vecTy = dyn_cast<VectorType>(getType())) {
      return SplatElementsAttr::get(vecTy, tureAttr);
    }
```

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 2]
```diff
@@ -309,19 +309,62 @@ OpFoldResult spirv::LogicalAndOp::fold(FoldAdaptor adaptor) {
...
+  }
+
+  return constFoldBinaryOp<IntegerAttr>(adaptor.getOperands(),
+                                        [](const APInt &a, const APInt &b) {
+                                          APInt zero = APInt::getZero(1);
+                                          return a == b ? (zero + 1) : zero;
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:334]: nit:
```suggestion
                                        [](const APInt &a, const APInt &b) {
                                          return a == b ? APInt::getAllOnes(1) : APInt::getZero(1);
```

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 3]
```diff
@@ -309,19 +309,62 @@ OpFoldResult spirv::LogicalAndOp::fold(FoldAdaptor adaptor) {
...
+  if (getOperand1() == getOperand2()) {
+    auto type = getType();
+    if (isa<IntegerType>(type)) {
+      return BoolAttr::get(getContext(), false);
+    }
+    if (isa<VectorType>(type)) {
+      auto vtType = cast<ShapedType>(type);
+      auto element = BoolAttr::get(getContext(), false);
+      return DenseElementsAttr::get(vtType, element);
+    }
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:360]: same here

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 4]
```diff
@@ -309,19 +309,62 @@ OpFoldResult spirv::LogicalAndOp::fold(FoldAdaptor adaptor) {
...
+  }
+
+  return constFoldBinaryOp<IntegerAttr>(adaptor.getOperands(),
+                                        [](const APInt &a, const APInt &b) {
+                                          APInt zero = APInt::getZero(1);
+                                          return a == b ? zero : (zero + 1);
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:366]: same here

**Review by kuhar — COMMENTED** [review]
Thanks, just some coding style suggestions here

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp (hunk 1 above)
Comment (inbelic) [reply mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp]: Thanks, that looks better

**Comment by inbelic** [comment]
Nice, thanks. Let me know if it looks good and I can rebase and try my new commit privileges :)

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 5]
```diff
@@ -662,19 +662,56 @@ OpFoldResult spirv::LogicalAndOp::fold(FoldAdaptor adaptor) {
...
+//===----------------------------------------------------------------------===//
+
+OpFoldResult
+spirv::LogicalEqualOp::fold(spirv::LogicalEqualOp::FoldAdaptor adaptor) {
+  // x == x -> true
+  if (getOperand1() == getOperand2()) {
+    auto trueAttr = BoolAttr::get(getContext(), true);
+    if (isa<IntegerType>(getType())) {
+      return trueAttr;
+    }
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:676]: nit: we don't need braces around single-line if statements. also elsewhere

**Review by kuhar — APPROVED** [review]
LGTM, just one cosmetic suggestion for single-line if statements. Feel free to change and merge without asking for re-approval.

**Pull request merged**